File: src/test/java/com/joecollins/graphics/components/SwingFrameBuilderTest.java
package com.joecollins.graphics.components;

import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.Binding;
import com.joecollins.graphics.utils.BindableWrapper;
import com.joecollins.models.general.Party;
import java.awt.Color;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.Test;

public class SwingFrameBuilderTest {

  private static class SwingProperties {
    private final Color leftColor;
    private final Color rightColor;
    private final Number value;
    private final String text;

    private SwingProperties(Color leftColor, Color rightColor, Number value, String text) {
      this.leftColor = leftColor;
      this.rightColor = rightColor;
      this.value = value;
      this.text = text;
    }
  }

  @Test
  public void basicTest() {
    BindableWrapper<SwingProperties> swingProps =
        new BindableWrapper<>(new SwingProperties(Color.RED, Color.BLUE, 0.02, "2% SWING"));
    SwingFrame frame =
        SwingFrameBuilder.basic(
                swingProps.getBinding(),
                p -> p.leftColor,
                p -> p.rightColor,
                p -> p.value,
                p -> p.text)
            .withRange(Binding.fixedBinding(0.10))
            .withHeader(Binding.fixedBinding("SWING"))
            .build();
    assertEquals(Color.RED, frame.getLeftColor());
    assertEquals(Color.BLUE, frame.getRightColor());
    assertEquals(Color.RED, frame.getBottomColor());
    assertEquals(0.02, frame.getValue().doubleValue(), 1e-6);
    assertEquals("2% SWING", frame.getBottomText());
    assertEquals(0.10, frame.getRange());
    assertEquals("SWING", frame.getHeader());

    swingProps.setValue(new SwingProperties(Color.GREEN, Color.ORANGE, -0.05, "5% SWING"));
    assertEquals(Color.GREEN, frame.getLeftColor());
    assertEquals(Color.ORANGE, frame.getRightColor());
    assertEquals(Color.ORANGE, frame.getBottomColor());
    assertEquals(-0.05, frame.getValue().doubleValue(), 1e-6);
    assertEquals("5% SWING", frame.getBottomText());
  }

  @Test
  public void testNeutralBottomColor() {
    BindableWrapper<SwingProperties> swingProps =
        new BindableWrapper<>(new SwingProperties(Color.RED, Color.BLUE, 0.02, "2% SWING"));
    BindableWrapper<Color> neutralColor = new BindableWrapper<>(Color.GRAY);
    SwingFrame frame =
        SwingFrameBuilder.basic(
                swingProps.getBinding(),
                p -> p.leftColor,
                p -> p.rightColor,
                p -> p.value,
                p -> p.text)
            .withRange(Binding.fixedBinding(0.10))
            .withNeutralColor(neutralColor.getBinding())
            .build();
    assertEquals(Color.RED, frame.getBottomColor());

    swingProps.setValue(new SwingProperties(Color.GREEN, Color.ORANGE, -0.05, "5% SWING"));
    assertEquals(Color.ORANGE, frame.getBottomColor());

    neutralColor.setValue(Color.LIGHT_GRAY);
    assertEquals(Color.ORANGE, frame.getBottomColor());

    swingProps.setValue(new SwingProperties(Color.GREEN, Color.ORANGE, 0.00, "NO SWING"));
    assertEquals(Color.LIGHT_GRAY, frame.getBottomColor());

    neutralColor.setValue(Color.BLACK);
    assertEquals(Color.BLACK, frame.getBottomColor());
  }

  @Test
  public void testSwingPrevCurrTwoMainPartiesSwingRight() {
    Party lib = new Party("LIBERAL", "LIB", Color.RED);
    Party con = new Party("CONSERVATIVE", "CON", Color.BLUE);
    Party ndp = new Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE);
    Binding<Map<Party, Integer>> prevBinding = () -> Map.of(lib, 25, con, 15, ndp, 10);
    Binding<Map<Party, Integer>> currBinding = () -> Map.of(lib, 16, con, 13, ndp, 11);
    // LIB: 50.00 -> 40.00 (-10.00)
    // CON: 30.00 -> 32.25 (+ 2.25)
    // NDP: 20.00 -> 27.75 (+ 7.75)
    List<Party> partyOrder = Arrays.asList(ndp, lib, con);
    SwingFrame swingFrame =
        SwingFrameBuilder.prevCurr(
                prevBinding, currBinding, Comparator.comparing(partyOrder::indexOf))
            .build();
    assertEquals(Color.BLUE, swingFrame.getLeftColor());
    assertEquals(Color.RED, swingFrame.getRightColor());
    assertEquals(Color.BLUE, swingFrame.getBottomColor());
    assertEquals(0.0625, swingFrame.getValue().doubleValue(), 1e-6);
    assertEquals(0.1, swingFrame.getRange().doubleValue(), 1e-6);
    assertEquals("6.2% SWING LIB TO CON", swingFrame.getBottomText());
  }

  @Test
  public void testSwingPrevCurrTwoMainPartiesSwingLeft() {
    Party lib = new Party("LIBERAL", "LIB", Color.RED);
    Party con = new Party("CONSERVATIVE", "CON", Color.BLUE);
    Party ndp = new Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE);
    Binding<Map<Party, Integer>> prevBinding = () -> Map.of(lib, 25, con, 15, ndp, 10);
    Binding<Map<Party, Integer>> currBinding = () -> Map.of(lib, 26, con, 10, ndp, 4);
    // LIB: 50.00 -> 65.00 (+15.00)
    // CON: 30.00 -> 25.00 (- 5.00)
    // NDP: 20.00 -> 10.00 (-10.00)
    List<Party> partyOrder = Arrays.asList(ndp, lib, con);
    SwingFrame swingFrame =
        SwingFrameBuilder.prevCurr(
                prevBinding, currBinding, Comparator.comparing(partyOrder::indexOf))
            .build();
    assertEquals(Color.BLUE, swingFrame.getLeftColor());
    assertEquals(Color.RED, swingFrame.getRightColor());
    assertEquals(Color.RED, swingFrame.getBottomColor());
    assertEquals(-0.1, swingFrame.getValue().doubleValue(), 1e-6);
    assertEquals(0.1, swingFrame.getRange().doubleValue(), 1e-6);
    assertEquals("10.0% SWING CON TO LIB", swingFrame.getBottomText());
  }

  @Test
  public void testSwingPrevCurrPartiesNotInComparator() {
    Party lib = new Party("LIBERAL", "LIB", Color.RED);
    Party con = new Party("CONSERVATIVE", "CON", Color.BLUE);
    Party ndp = new Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE);
    Binding<Map<Party, Integer>> prevBinding = () -> Map.of(lib, 25, con, 15, ndp, 10);
    Binding<Map<Party, Integer>> currBinding = () -> Map.of(lib, 26, con, 10, ndp, 4);
    // LIB: 50.00 -> 65.00 (+15.00)
    // CON: 30.00 -> 25.00 (- 5.00)
    // NDP: 20.00 -> 10.00 (-10.00)
    List<Party> partyOrder = Arrays.asList();
    SwingFrame swingFrame =
        SwingFrameBuilder.prevCurr(
                prevBinding, currBinding, Comparator.comparing(partyOrder::indexOf))
            .build();
    assertEquals(
        Set.of(Color.BLUE, Color.RED),
        Set.of(swingFrame.getLeftColor(), swingFrame.getRightColor()));
    //    assertEquals(Color.RED, swingFrame.getBottomColor());
    //    assertEquals(0.1 * (swingFrame.getLeftColor().equals(Color.BLUE) ? -1 : 1),
    // swingFrame.getValue().doubleValue(), 1e-6);
    assertEquals(0.1, swingFrame.getRange().doubleValue(), 1e-6);
    assertEquals("10.0% SWING CON TO LIB", swingFrame.getBottomText());
  }

  @Test
  public void testSwingPrevCurrPartiesSwingLeftFromRight() {
    Party lib = new Party("LIBERAL", "LIB", Color.RED);
    Party con = new Party("CONSERVATIVE", "CON", Color.BLUE);
    Party ndp = new Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE);
    Binding<Map<Party, Integer>> prevBinding = () -> Map.of(lib, 15, con, 25, ndp, 10);
    Binding<Map<Party, Integer>> currBinding = () -> Map.of(lib, 6, con, 10, ndp, 24);
    // LIB: 30.00 -> 15.00 (-15.00)
    // CON: 50.00 -> 25.00 (-25.00)
    // NDP: 20.00 -> 60.00 (+40.00)
    List<Party> partyOrder = Arrays.asList(ndp, lib, con);
    SwingFrame swingFrame =
        SwingFrameBuilder.prevCurr(
                prevBinding, currBinding, Comparator.comparing(partyOrder::indexOf))
            .build();
    assertEquals(Color.BLUE, swingFrame.getLeftColor());
    assertEquals(Color.ORANGE, swingFrame.getRightColor());
    assertEquals(Color.ORANGE, swingFrame.getBottomColor());
    assertEquals(-0.325, swingFrame.getValue().doubleValue(), 1e-6);
    assertEquals(0.1, swingFrame.getRange().doubleValue(), 1e-6);
    assertEquals("32.5% SWING CON TO NDP", swingFrame.getBottomText());
  }

  @Test
  public void testNoSwingBetweenParties() {
    Party lib = new Party("LIBERAL", "LIB", Color.RED);
    Party con = new Party("CONSERVATIVE", "CON", Color.BLUE);
    Party ndp = new Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE);
    Binding<Map<Party, Integer>> prevBinding = () -> Map.of(lib, 15, con, 25, ndp, 10);
    Binding<Map<Party, Integer>> currBinding = () -> Map.of(lib, 15, con, 25, ndp, 10);
    List<Party> partyOrder = Arrays.asList(ndp, lib, con);
    SwingFrame swingFrame =
        SwingFrameBuilder.prevCurr(
                prevBinding, currBinding, Comparator.comparing(partyOrder::indexOf))
            .build();
    assertEquals(Color.BLUE, swingFrame.getLeftColor());
    assertEquals(Color.RED, swingFrame.getRightColor());
    assertEquals(Color.LIGHT_GRAY, swingFrame.getBottomColor());
    assertEquals(0.0, swingFrame.getValue().doubleValue(), 1e-6);
    assertEquals(0.1, swingFrame.getRange().doubleValue(), 1e-6);
    assertEquals("NO SWING", swingFrame.getBottomText());
  }

  @Test
  public void testNoSwingAvailable() {
    Party lib = new Party("LIBERAL", "LIB", Color.RED);
    Party con = new Party("CONSERVATIVE", "CON", Color.BLUE);
    Party ndp = new Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE);
    Binding<Map<Party, Integer>> prevBinding = () -> Map.of(lib, 15, con, 25, ndp, 10);
    Binding<Map<Party, Integer>> currBinding = Map::of;
    List<Party> partyOrder = Arrays.asList(ndp, lib, con);
    SwingFrame swingFrame =
        SwingFrameBuilder.prevCurr(
                prevBinding, currBinding, Comparator.comparing(partyOrder::indexOf))
            .build();
    assertEquals(Color.LIGHT_GRAY, swingFrame.getLeftColor());
    assertEquals(Color.LIGHT_GRAY, swingFrame.getRightColor());
    assertEquals(Color.LIGHT_GRAY, swingFrame.getBottomColor());
    assertEquals(0.0, swingFrame.getValue().doubleValue(), 1e-6);
    assertEquals(0.1, swingFrame.getRange().doubleValue(), 1e-6);
    assertEquals("NOT AVAILABLE", swingFrame.getBottomText());
  }

  @Test
  public void testSwingPrevCurrTwoMainPartiesNormalised() {
    Party lib = new Party("LIBERAL", "LIB", Color.RED);
    Party con = new Party("CONSERVATIVE", "CON", Color.BLUE);
    Party ndp = new Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE);
    Binding<Map<Party, Double>> prevBinding = () -> Map.of(lib, 0.40, con, 0.30, ndp, 0.20);
    Binding<Map<Party, Double>> currBinding = () -> Map.of(lib, 0.38, con, 0.35, ndp, 0.18);
    // LIB: 40.00 -> 38.00 (- 2.00)
    // CON: 30.00 -> 35.00 (+ 5.00)
    // NDP: 20.00 -> 18.00 (- 2.00)
    List<Party> partyOrder = Arrays.asList(ndp, lib, con);
    SwingFrame swingFrame =
        SwingFrameBuilder.prevCurrNormalised(
                prevBinding, currBinding, Comparator.comparing(partyOrder::indexOf))
            .build();
    assertEquals(Color.BLUE, swingFrame.getLeftColor());
    assertEquals(Color.RED, swingFrame.getRightColor());
    assertEquals(Color.BLUE, swingFrame.getBottomColor());
    assertEquals(0.035, swingFrame.getValue().doubleValue(), 1e-6);
    assertEquals(0.1, swingFrame.getRange().doubleValue(), 1e-6);
    assertEquals("3.5% SWING LIB TO CON", swingFrame.getBottomText());
  }
}

File: src/test/kotlin/com/joecollins/graphics/components/SwingFrameBuilderTest.kt
package com.joecollins.graphics.components

import com.joecollins.graphics.components.SwingFrameBuilder.Companion.basic
import com.joecollins.graphics.components.SwingFrameBuilder.Companion.prevCurr
import com.joecollins.graphics.components.SwingFrameBuilder.Companion.prevCurrNormalised
import com.joecollins.graphics.utils.ColorUtils
import com.joecollins.models.general.Party
import com.joecollins.pubsub.Publisher
import com.joecollins.pubsub.asOneTimePublisher
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.Assert
import org.junit.Test
import java.awt.Color
import java.util.concurrent.TimeUnit

class SwingFrameBuilderTest {
    private class SwingProperties(val leftColor: Color, val rightColor: Color, val value: Number, val text: String)

    @Test
    fun basicTest() {
        val swingProps = Publisher(SwingProperties(Color.RED, Color.BLUE, 0.02, "2% SWING"))
        val frame = basic(
            swingProps,
            { p: SwingProperties -> p.leftColor },
            { p: SwingProperties -> p.rightColor },
            { p: SwingProperties -> p.value },
            { p: SwingProperties -> p.text }
        )
            .withRange(0.10.asOneTimePublisher())
            .withHeader("SWING".asOneTimePublisher())
            .build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getLeftColor() }, IsEqual(Color.RED))
        Assert.assertEquals(Color.BLUE, frame.getRightColor())
        Assert.assertEquals(Color.RED, frame.getBottomColor())
        Assert.assertEquals(0.02, frame.getValue().toDouble(), 1e-6)
        Assert.assertEquals("2% SWING", frame.getBottomText())
        Assert.assertEquals(0.10, frame.getRange())
        Assert.assertEquals("SWING", frame.header)
        swingProps.submit(SwingProperties(Color.GREEN, Color.ORANGE, -0.05, "5% SWING"))
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getLeftColor() }, IsEqual(Color.GREEN))
        Assert.assertEquals(Color.ORANGE, frame.getRightColor())
        Assert.assertEquals(ColorUtils.contrastForBackground(Color.ORANGE), frame.getBottomColor())
        Assert.assertEquals(-0.05, frame.getValue().toDouble(), 1e-6)
        Assert.assertEquals("5% SWING", frame.getBottomText())
    }

    @Test
    fun testNeutralBottomColor() {
        val swingProps = Publisher(SwingProperties(Color.RED, Color.BLUE, 0.02, "2% SWING"))
        val neutralColor = Publisher(Color.GRAY)
        val frame = basic(
            swingProps,
            { p: SwingProperties -> p.leftColor },
            { p: SwingProperties -> p.rightColor },
            { p: SwingProperties -> p.value },
            { p: SwingProperties -> p.text }
        )
            .withRange(0.10.asOneTimePublisher())
            .withNeutralColor(neutralColor)
            .build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getBottomColor() }, IsEqual(Color.RED))
        swingProps.submit(SwingProperties(Color.GREEN, Color.ORANGE, -0.05, "5% SWING"))
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getBottomColor() }, IsEqual(ColorUtils.contrastForBackground(Color.ORANGE)))
        neutralColor.submit(Color.LIGHT_GRAY)
        Awaitility.await().atMost(600, TimeUnit.MILLISECONDS).pollDelay(500, TimeUnit.MILLISECONDS)
            .until({ frame.getBottomColor() }, IsEqual(ColorUtils.contrastForBackground(Color.ORANGE)))
        swingProps.submit(SwingProperties(Color.GREEN, Color.ORANGE, 0.00, "NO SWING"))
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getBottomColor() }, IsEqual(ColorUtils.contrastForBackground(Color.LIGHT_GRAY)))
        neutralColor.submit(Color.BLACK)
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getBottomColor() }, IsEqual(Color.BLACK))
    }

    @Test
    fun testSwingPrevCurrTwoMainPartiesSwingRight() {
        val lib = Party("LIBERAL", "LIB", Color.RED)
        val con = Party("CONSERVATIVE", "CON", Color.BLUE)
        val ndp = Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE)
        val prevBinding = mapOf(lib to 25, con to 15, ndp to 10).asOneTimePublisher()
        val currBinding = mapOf(lib to 16, con to 13, ndp to 11).asOneTimePublisher()
        // LIB: 50.00 -> 40.00 (-10.00)
        // CON: 30.00 -> 32.25 (+ 2.25)
        // NDP: 20.00 -> 27.75 (+ 7.75)
        val partyOrder = listOf(ndp, lib, con)
        val swingFrame = prevCurr(
            prevBinding, currBinding, compareBy { partyOrder.indexOf(it) }
        )
            .build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ swingFrame.getLeftColor() }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.RED, swingFrame.getRightColor())
        Assert.assertEquals(Color.BLUE, swingFrame.getBottomColor())
        Assert.assertEquals(0.0625, swingFrame.getValue().toDouble(), 1e-6)
        Assert.assertEquals(0.1, swingFrame.getRange().toDouble(), 1e-6)
        Assert.assertEquals("6.2% SWING LIB TO CON", swingFrame.getBottomText())
    }

    @Test
    fun testSwingPrevCurrTwoMainPartiesSwingLeft() {
        val lib = Party("LIBERAL", "LIB", Color.RED)
        val con = Party("CONSERVATIVE", "CON", Color.BLUE)
        val ndp = Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE)
        val prevBinding = mapOf(lib to 25, con to 15, ndp to 10).asOneTimePublisher()
        val currBinding = mapOf(lib to 26, con to 10, ndp to 4).asOneTimePublisher()
        // LIB: 50.00 -> 65.00 (+15.00)
        // CON: 30.00 -> 25.00 (- 5.00)
        // NDP: 20.00 -> 10.00 (-10.00)
        val partyOrder = listOf(ndp, lib, con)
        val swingFrame = prevCurr(prevBinding, currBinding, compareBy { partyOrder.indexOf(it) }).build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ swingFrame.getLeftColor() }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.RED, swingFrame.getRightColor())
        Assert.assertEquals(Color.RED, swingFrame.getBottomColor())
        Assert.assertEquals(-0.1, swingFrame.getValue().toDouble(), 1e-6)
        Assert.assertEquals(0.1, swingFrame.getRange().toDouble(), 1e-6)
        Assert.assertEquals("10.0% SWING CON TO LIB", swingFrame.getBottomText())
    }

    @Test
    fun testSwingPrevCurrPartiesNotInComparator() {
        val lib = Party("LIBERAL", "LIB", Color.RED)
        val con = Party("CONSERVATIVE", "CON", Color.BLUE)
        val ndp = Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE)
        val prevBinding = mapOf(lib to 25, con to 15, ndp to 10).asOneTimePublisher()
        val currBinding = mapOf(lib to 26, con to 10, ndp to 4).asOneTimePublisher()
        // LIB: 50.00 -> 65.00 (+15.00)
        // CON: 30.00 -> 25.00 (- 5.00)
        // NDP: 20.00 -> 10.00 (-10.00)
        val partyOrder: List<Party> = listOf()
        val swingFrame = prevCurr(
            prevBinding, currBinding, compareBy { partyOrder.indexOf(it) }
        )
            .build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ swingFrame.getBottomText() }, IsEqual("10.0% SWING CON TO LIB"))
        Assert.assertEquals(
            setOf(Color.BLUE, Color.RED),
            setOf(swingFrame.getLeftColor(), swingFrame.getRightColor())
        )
        //    assertEquals(Color.RED, swingFrame.getBottomColor());
        //    assertEquals(0.1 * (swingFrame.getLeftColor().equals(Color.BLUE) ? -1 : 1),
        // swingFrame.getValue().doubleValue(), 1e-6);
        Assert.assertEquals(0.1, swingFrame.getRange().toDouble(), 1e-6)
        Assert.assertEquals("10.0% SWING CON TO LIB", swingFrame.getBottomText())
    }

    @Test
    fun testSwingPrevCurrPartiesSwingLeftFromRight() {
        val lib = Party("LIBERAL", "LIB", Color.RED)
        val con = Party("CONSERVATIVE", "CON", Color.BLUE)
        val ndp = Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE)
        val prevBinding = mapOf(lib to 15, con to 25, ndp to 10).asOneTimePublisher()
        val currBinding = mapOf(lib to 6, con to 10, ndp to 24).asOneTimePublisher()
        // LIB: 30.00 -> 15.00 (-15.00)
        // CON: 50.00 -> 25.00 (-25.00)
        // NDP: 20.00 -> 60.00 (+40.00)
        val partyOrder = listOf(ndp, lib, con)
        val swingFrame = prevCurr(prevBinding, currBinding, compareBy { partyOrder.indexOf(it) }).build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ swingFrame.getLeftColor() }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.ORANGE, swingFrame.getRightColor())
        Assert.assertEquals(ColorUtils.contrastForBackground(Color.ORANGE), swingFrame.getBottomColor())
        Assert.assertEquals(-0.325, swingFrame.getValue().toDouble(), 1e-6)
        Assert.assertEquals(0.1, swingFrame.getRange().toDouble(), 1e-6)
        Assert.assertEquals("32.5% SWING CON TO NDP", swingFrame.getBottomText())
    }

    @Test
    fun testNoSwingBetweenParties() {
        val lib = Party("LIBERAL", "LIB", Color.RED)
        val con = Party("CONSERVATIVE", "CON", Color.BLUE)
        val ndp = Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE)
        val prevBinding = mapOf(lib to 15, con to 25, ndp to 10).asOneTimePublisher()
        val currBinding = mapOf(lib to 15, con to 25, ndp to 10).asOneTimePublisher()
        val partyOrder = listOf(ndp, lib, con)
        val swingFrame = prevCurr(prevBinding, currBinding, Comparator.comparing { o: Party -> partyOrder.indexOf(o) }).build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ swingFrame.getLeftColor() }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.RED, swingFrame.getRightColor())
        Assert.assertEquals(ColorUtils.contrastForBackground(Color.LIGHT_GRAY), swingFrame.getBottomColor())
        Assert.assertEquals(0.0, swingFrame.getValue().toDouble(), 1e-6)
        Assert.assertEquals(0.1, swingFrame.getRange().toDouble(), 1e-6)
        Assert.assertEquals("NO SWING", swingFrame.getBottomText())
    }

    @Test
    fun testNoSwingAvailable() {
        val lib = Party("LIBERAL", "LIB", Color.RED)
        val con = Party("CONSERVATIVE", "CON", Color.BLUE)
        val ndp = Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE)
        val prevBinding = mapOf(lib to 15, con to 25, ndp to 10).asOneTimePublisher()
        val currBinding = emptyMap<Party, Int>().asOneTimePublisher()
        val partyOrder = listOf(ndp, lib, con)
        val swingFrame = prevCurr(prevBinding, currBinding, Comparator.comparing { o: Party -> partyOrder.indexOf(o) }).build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ swingFrame.getLeftColor() }, IsEqual(Color.LIGHT_GRAY))
        Assert.assertEquals(Color.LIGHT_GRAY, swingFrame.getRightColor())
        Assert.assertEquals(ColorUtils.contrastForBackground(Color.LIGHT_GRAY), swingFrame.getBottomColor())
        Assert.assertEquals(0.0, swingFrame.getValue().toDouble(), 1e-6)
        Assert.assertEquals(0.1, swingFrame.getRange().toDouble(), 1e-6)
        Assert.assertEquals("NOT AVAILABLE", swingFrame.getBottomText())
    }

    @Test
    fun testSwingPrevCurrTwoMainPartiesNormalised() {
        val lib = Party("LIBERAL", "LIB", Color.RED)
        val con = Party("CONSERVATIVE", "CON", Color.BLUE)
        val ndp = Party("NEW DEMOCRATIC PARTY", "NDP", Color.ORANGE)
        val prevBinding = mapOf(lib to 0.40, con to 0.30, ndp to 0.20).asOneTimePublisher()
        val currBinding = mapOf(lib to 0.38, con to 0.35, ndp to 0.18).asOneTimePublisher()
        // LIB: 40.00 -> 38.00 (- 2.00)
        // CON: 30.00 -> 35.00 (+ 5.00)
        // NDP: 20.00 -> 18.00 (- 2.00)
        val partyOrder = listOf(ndp, lib, con)
        val swingFrame = prevCurrNormalised(prevBinding, currBinding, Comparator.comparing { o: Party -> partyOrder.indexOf(o) }).build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ swingFrame.getLeftColor() }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.RED, swingFrame.getRightColor())
        Assert.assertEquals(Color.BLUE, swingFrame.getBottomColor())
        Assert.assertEquals(0.035, swingFrame.getValue().toDouble(), 1e-6)
        Assert.assertEquals(0.1, swingFrame.getRange().toDouble(), 1e-6)
        Assert.assertEquals("3.5% SWING LIB TO CON", swingFrame.getBottomText())
    }
}
